src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,129 @@
+        addEntry(hash, key, value, index);
     }
 
     @Override
     public V getValue(K key) {
-        return null;
+        if (key == null) {
+            return getForNullKey();
+        }
+
+        int hash = hash(key);
+        int index = indexFor(hash, table.length);
+
+        for (Entry<K, V> entry = table[index]; entry != null; entry = entry.next) {
+            if ((entry.hash == hash) && (key == entry.key || key.equals(entry.key))) {
+                return entry.value; // Return the value if key is found
+            }
+        }
+
+        return null; // Return null if key is not found
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    // method to handle null key
+    private void putForNullKey(V value) {
+        for (Entry<K, V> entry = table[0]; entry != null; entry = entry.next) {
+            if (entry.key == null) {
+                entry.value = value; // Update the value if null key already exists
+                return;
+            }
+        }
+        addEntry(0, null, value, 0);
+    }
+
+    // method to handle null key
+    private V getForNullKey() {
+        for (Entry<K, V> entry = table[0]; entry != null; entry = entry.next) {
+            if (entry.key == null) {
+                return entry.value; // Return the value if null key is found
+            }
+        }
+        return null; // Return null if null key is not found
+    }
+
+    // method to calculate hash
+    private int hash(Object key) {
+        return key == null ? 0 : key.hashCode();
+    }
+
+    // method to calculate index for the given hash and table length
+    private int indexFor(int hash, int length) {
+        return hash & (length - 1);
+    }
+
+    // method to add a new entry to the table
+    private void addEntry(int hash, K key, V value, int index) {
+        Entry<K, V> entry = table[index];
+        table[index] = new Entry<>(hash, key, value, entry);
+        size++;
+
+        // Check if resizing is needed
+        if (size > table.length * DEFAULT_LOAD_FACTOR) {
+            resize();
+        }
+    }
+
+    private void resize() {
+        int newCapacity = table.length * RESIZE_FACTOR;
+        Entry<K, V>[] newTable = new Entry[newCapacity];
+
+        for (Entry<K, V> entry : table) {
+            while (entry != null) {
+                Entry<K, V> next = entry.next;
+                int index = indexFor(entry.hash, newCapacity);
+                entry.next = newTable[index];
+                newTable[index] = entry;
+                entry = next;
+            }
+        }
+        table = newTable;
+    }
+
+    static class Entry<K, V> {
+        private final int hash;
+        private final K key;
+        private V value;
+        private Entry<K, V> next;
+
+        Entry(int hash, K key, V value, Entry<K, V> next) {
+            this.hash = hash;
+            this.key = key;
+            this.value = value;
+            this.next = next;
```

Comment:
hash is a calculable field
```suggestion
        private final K key;
        private V value;
        private Entry<K, V> next;

        private Entry(K key, V value, Entry<K, V> next) {
            this.key = key;
            this.value = value;
            this.next = next;
```